### PR TITLE
Add support for iterating only loaded values

### LINKED ISSUE #303: Iteration over loaded values
### Issue To Be Solved

In Cadence, it would be useful to have a mechanism to iterate over an atree array or ordered map, which only considers slabs and values which have been previously loaded from storage.

For example, if a large array or map is loaded from storage, only its root slab is loaded, and none of the elements (child slabs) are loaded.
If only one element is accessed, then only the slabs in the tree leading to the element is loaded into memory from storage, other slabs and elements are not.

### Suggested Solution

I do not have a full solution in mind, but investigated a bit how this could be achieved.

For example, in the array iterator:
- When retrieving a slab, the iterator should do so only if the slab had been previously retrieved (https://github.com/onflow/atree/blob/bc0184e2e9ba33d9d30aec2168dba35ccc62982d/array.go#L2202). This probably needs an additional function in `SlabStorage`, or an additional parameter for the `Retrieve` function to indicate if the value should be retrieved if it has not been previously retrieved (default).
- When iterating over elements, the iterator should only load values if they have been previously loaded (https://github.com/onflow/atree/blob/bc0184e2e9ba33d9d30aec2168dba35ccc62982d/array.go#L2218). This probably needs an alternative function in `Storable`, or an additional parameter for the `StoredValue` (like for `Retrieve`).

Would it be possible to add such a feature? Would such an approach work?

### PR DESCRIPTION
**Pull request by fxamacker**
Closes #303

## Description

Added new functions and structs for a special iterators required by Stable Cadence:
- Array.IterateLoadedValues()
- ArrayLoadedValueIterator
- OrderedMap.IterateLoadedValues()
- MapLoadedValueIterator etc.

They have same API as regular iterators except that they only return already loaded elements.

______

- [x] Targeted PR against `main` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/atree/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: array.go [hunk 1]
```diff
@@ -2709,3 +2709,157 @@ func nextLevelArraySlabs(storage SlabStorage, address Address, slabs []ArraySlab
...
+			// Wrap err as external error (if needed) because err is returned by ArrayIterationFunc callback.
+			return wrapErrorAsExternalErrorIfNeeded(err)
+		}
+		if !resume {
+			return nil
+		}
+	}
+}
+
+func (a *Array) getLoadedDataSlabs() ([]*ArrayDataSlab, error) {
```
Comment (turbolent) [inline array.go:2824]: Would it be possible to perform this during the iteration?

Currently all loaded slabs are searched at the beginning of iteration, and then the loaded slabs are iterated over. This might become very costly for large arrays.

Maybe we can use the code of this function in the iterator's next function, and persist intermediate state which is currently kept on the stack, inside of the iterator instead.

**Review by turbolent — COMMENTED** [review]
Great work!

File: map.go [hunk 2]
```diff
@@ -4626,3 +4626,281 @@ func nextLevelMapSlabs(storage SlabStorage, address Address, slabs []MapSlab) ([
...
+			newParent := &mapLoadedSlabIterator{
+				storage: i.storage,
+				slab:    slab,
+			}
+			i.parents = append(i.parents, newParent)
+
+		case nil:
+			// Reach end of last parent.
+			// Pop last parent from parents stack.
+			i.parents = i.parents[:len(i.parents)-1]
```
Comment (turbolent) [inline map.go:4797]: Same as above

**Review by turbolent — APPROVED** [review]
Really great work!

**Comment by turbolent** [comment]
Maybe Supun (once he's back), Ramtin (once he's back), and/or someone else from the execution team can have a look as well before we merge

File: array.go (hunk 1 above)
Comment (turbolent) [reply array.go]: Great refactor and improvement! 👏

**Comment by turbolent** [comment]
@SupunS could you please also have a look here?

**Comment by ramtinms** [comment]
I might need some context on this one. 
- Who would be the final user of this feature? cadence runtime code and or would it be directly exposed to the user's scripts? 
- How often the Atree internal cache of slabs is purged? I assume is at least once per transaction otherwise we might return values without reading them from the storage and cause missing register errors.

**Comment by fxamacker** [comment]
@ramtinms Thanks for taking a look!

> I might need some context on this one.
> Who would be the final user of this feature? cadence runtime code and or would it be directly exposed to the user's scripts?

I think the plan is Cadence runtime.  Bastian and Supun provided some context in reply to [this comment](https://github.com/onflow/atree/issues/303#issuecomment-1540523018) on the issue.

> How often the Atree internal cache of slabs is purged? I assume is at least once per transaction otherwise we might return values without reading them from the storage and cause missing register errors.

I think internal cache is purged on commit but I'm not sure how often commit is called. Maybe @turbolent or @janezpodhostnik can provide more info about how often that happens.

**Review by SupunS — APPROVED** [review]
Thank you for adding this!

As usual, I don't know much about the atree internals, so mostly reviewed the general Go code. Looks great!

**Review by ramtinms — APPROVED** [review]
Looks good to me, 
I only recommend adding a small test that changes the map or array contents in the middle of an iteration. not sure this would happen ever but that's the only edge case that comes to my mind that didn't find it tests.

**Review by janezpodhostnik — APPROVED** [review]
Code looks good to me. I was especially looking for any sources of non-determinism, and it looks all good 👍.
